xarray/core/indexes.py

```diff
@@ -25,17 +25,60 @@
 from .utils import Frozen, get_valid_numpy_dtype, is_dict_like, is_scalar
 
 if TYPE_CHECKING:
-    from .types import ErrorOptions, T_Index
+    from .types import ErrorOptions, JoinOptions, T_Index
     from .variable import Variable
 
 IndexVars = Dict[Any, "Variable"]
 
 
 class Index:
-    """Base class inherited by all xarray-compatible indexes."""
+    """Base class inherited by all xarray-compatible indexes.
+
+    Do not use this class directly for creating index objects. Xarray indexes
+    are created exclusively from subclasses of ``Index``, mostly via Xarray's
+    public API like ``Dataset.set_xindex``.
+
+    Every subclass must at least implement :py:meth:`Index.from_variables`. The
+    (re)implementation of the other methods of this base class is optional but
+    mostly required in order to support operations relying on indexes such as
+    label-based selection or alignment.
+
+    The ``Index`` API closely follows the :py:meth:`Dataset` and
+    :py:meth:`DataArray` API, e.g., for an index to support ``.sel()`` it needs
+    to implement :py:meth:`Index.sel`, to support ``.stack()`` and
+    ``.unstack()`` it needs to implement :py:meth:`Index.stack` and
+    :py:meth:`Index.unstack`, etc.
+
+    When a method is not (re)implemented, depending on the case the
+    corresponding operation on a :py:meth:`Dataset` or :py:meth:`DataArray`
+    either will raise a ``NotImplementedError`` or will simply drop/pass/copy
+    the index from/to the result.
+
+    """
 
     @classmethod
-    def from_variables(cls, variables: Mapping[Any, Variable]) -> Index:
+    def from_variables(
```

Comment:
Tangential but couldn't this be decorated to make it an abstract (class) method? Then an error would be raised if subclasses don't implement it. i.e. this:

```python
from abc import ABC, abstractmethod

class Index(ABC):
     @classmethod
     @abstractmethod
     def from_variables(cls, vars):
         ...

     def sel(self, indexers):
         raise NotImplementedError()
 
```